azure/hostedservice/client.go

```diff
@@ -0,0 +1,135 @@
+	if len(dnsName) == 0 {
+		return "", fmt.Errorf(errParamNotSpecified, "dnsName")
+	}
+	if len(location) == 0 {
+		return "", fmt.Errorf(errParamNotSpecified, "location")
+	}
+
+	err := self.verifyDNSName(dnsName)
+	if err != nil {
+		return "", err
+	}
+
+	result, reason, err := self.CheckHostedServiceNameAvailability(dnsName)
+	if err != nil {
+		return "", err
+	}
+	if !result {
+		return "", fmt.Errorf("%s Hosted service name: %s", reason, dnsName)
+	}
+
+	locationClient := locationclient.NewClient(self.client)
+	err = locationClient.ResolveLocation(location)
+	if err != nil {
+		return "", err
+	}
+
+	hostedServiceDeployment := self.createHostedServiceDeploymentConfig(dnsName, location, reverseDnsFqdn, label, description)
+	hostedServiceBytes, err := xml.Marshal(hostedServiceDeployment)
+	if err != nil {
+		return "", err
+	}
+
+	requestURL := azureHostedServiceListURL
+	requestId, err := self.client.SendAzurePostRequest(requestURL, hostedServiceBytes)
+	if err != nil {
+		return "", err
+	}
+
+	return requestId, nil
+}
+
+func (self *HostedServiceClient) CheckHostedServiceNameAvailability(dnsName string) (bool, string, error) {
+	if len(dnsName) == 0 {
+		return false, "", fmt.Errorf(errParamNotSpecified, "dnsName")
+	}
+
+	err := self.verifyDNSName(dnsName)
+	if err != nil {
+		return false, "", err
+	}
+
+	requestURL := fmt.Sprintf(azureHostedServiceAvailabilityURL, dnsName)
+	response, err := self.client.SendAzureGetRequest(requestURL)
+	if err != nil {
+		return false, "", err
+	}
+
+	availabilityResponse := new(AvailabilityResponse)
+	err = xml.Unmarshal(response, availabilityResponse)
+	if err != nil {
+		return false, "", err
+	}
+
+	return availabilityResponse.Result, availabilityResponse.Reason, nil
+}
+
+func (self *HostedServiceClient) DeleteHostedService(dnsName string) error {
+	if len(dnsName) == 0 {
+		return fmt.Errorf(errParamNotSpecified, "dnsName")
+	}
+
+	err := self.verifyDNSName(dnsName)
+	if err != nil {
+		return err
+	}
+
+	requestURL := fmt.Sprintf(deleteAzureHostedServiceURL, dnsName)
+	requestId, err := self.client.SendAzureDeleteRequest(requestURL)
+	if err != nil {
+		return err
+	}
+
+	self.client.WaitAsyncOperation(requestId)
+	return nil
+}
+
+func (self *HostedServiceClient) createHostedServiceDeploymentConfig(dnsName, location string, reverseDnsFqdn string, label string, description string) HostedServiceDeployment {
+	deployment := HostedServiceDeployment{}
+	deployment.ServiceName = dnsName
+	encodedLabel := base64.StdEncoding.EncodeToString([]byte(label))
+	deployment.Label = encodedLabel
+	deployment.Description = description
+	deployment.Location = location
+	deployment.ReverseDnsFqdn = reverseDnsFqdn
+	deployment.Xmlns = azureXmlns
+	return deployment
+}
+
+func (self *HostedServiceClient) verifyDNSName(dns string) error {
+	if len(dns) < 3 || len(dns) > 25 {
```

Comment:
It's actually different from the restrictions the portal puts on them too - not sure which is correct actually. I just moved it across from the original code without investigating!